okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/FlowControlTest.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2018 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.remoting3.okhttp;
+
+import com.codahale.metrics.Meter;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.common.util.concurrent.RateLimiter;
+import com.netflix.concurrency.limits.Limiter;
+import com.palantir.remoting3.okhttp.ConcurrencyLimiters.ConcurrencyLimiter;
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is a simulation of the flow control primitives used by this library, in order to allow the developer
+ * to try different strategies.
+ * <p>
+ * It is run in CI, but only to prevent code breakages - this is in general a
+ */
+public final class FlowControlTest {
+    private static final Logger log = LoggerFactory.getLogger(FlowControlTest.class);
+    private static final int REQUESTS_PER_THREAD = System.getenv("CI") == null ? 1000 : 1;
+    private static final ConcurrencyLimiters limiters = new ConcurrencyLimiters();
+    private static ListeningExecutorService executorService;
+
+    private final ConcurrencyLimiter limiter = limiters.limiter(UUID.randomUUID().toString());
+
+    @BeforeClass
+    public static void beforeClass() {
+        executorService = MoreExecutors.listeningDecorator(Executors.newCachedThreadPool());
+    }
+
+    @AfterClass
+    public static void afterClass() {
+        executorService.shutdown();
+    }
+
+    @Test
+    public void test16ThreadsRateLimit20() throws ExecutionException, InterruptedException {
+        Meter rate = new Meter();
+        List<ListenableFuture<?>> tasks = createWorkers(rate, 16, 20, Duration.ofMillis(100))
```

Comment:
turns out the number of workers doesn't actually affect this until the rate limit is also large (because waiting isn't factored into this calculation).